### PR TITLE
Initial wording for SPIR binary clarification in clCompileProgram.

### LINKED ISSUE #659: `clCompileProgram` validity with a program created from SPIR
We're seeking clarification on the expected behaviour/steps required to compile and link a program created from a SPIR input (Note: **not** SPIR-**V** which is covered by wording regarding `clCreateProgramWithIL`)

We currently have these snippets of wording in the spec which we have interpreted to mean that calling `clCompileProgram` with a program object that was produced with `clCreateProgramWithBinary` where the supplied input was valid SPIR is an invalid operation:

> If program was created using clCreateProgramWithIL, then num_input_headers, input_headers, and header_include_names are ignored.

> CL_INVALID_OPERATION if program has no source or IL available, i.e. it has not been created with clCreateProgramWithSource or clCreateProgramWithIL.

However, our view is that a SPIR input _is_ valid IL and programs created with it via `clCreateProgramWithBinary` should be valid for use with `clCompileProgram`. If that explicitly by design, then is the expected behaviour that `clBuildProgram` must be called first, extracting binaries from that program and then re-creating a new program with those freshly extracted binaries?

There is a somewhat loosely related issue here: #520.

### PR DESCRIPTION
**Pull request by Jeremy-Kemp**
Fixes #659.

### CODE REVIEW

File: ext/cl_khr_spir.asciidoc [hunk 1]
```diff
@@ -90,6 +90,26 @@ For example, if the binary is as described in SPIR version 1.2, then
...
+"`If program was created using *clCreateProgramWithIL*, or *clCreateProgramWithBinary* with a corresponding `cl_program_binary_type` of `CL_PROGRAM_BINARY_TYPE_INTERMEDIATE`, then _num_input_headers_, _input_headers_,
+and _header_include_names_ are ignored.`"
+
+Modify following text in *clCompileProgram* from:
+
+"```CL_INVALID_OPERATION`` if program has no source or IL available, i.e. it has not been created with *clCreateProgramWithSource* or *clCreateProgramWithIL*.`"
+
+to:
+
+"``CL_INVALID_OPERATION`` if program has no source or IL available, i.e. it does not have a corresponding `cl_program_binary_type` of `CL_PROGRAM_BINARY_TYPE_NONE` or `CL_PROGRAM_BINARY_TYPE_INTERMEDIATE`."
```
[the remark this answers is not in the document]
Comment (Jeremy-Kemp) [reply ext/cl_khr_spir.asciidoc]: As discussed in the teleconference, this analysis is correct. Due to the state of a program changing after a call to `clCompileProgram`, this wording is incorrect.

I've pushed an update that moves the wording clarification to the core spec instead.

**Review by bashbaug — COMMENTED** [review]
The updated text looks good - thanks!

My only question is whether this should be described here in the main spec, or whether it should be described as a change to this error code in the extension spec for `cl_khr_spir`?

On one hand the _options_ are not specific to the SPIR extension, but on the other hand an implementation that does not support `cl_khr_spir` would never need to worry about this specific condition.

I don't have a very strong opinion about this, but I do have a very slight preference to describe it in the extension spec instead.  I'm interested what others think.

Thank you!

**Comment by alycm** [comment]
> My only question is whether this should be described here in the main spec, or whether it should be described as a change to this error code in the extension spec for `cl_khr_spir`?

I'd also have a mild preference for this going in the extension spec, SPIR is not currently mentioned in the API spec at all except in the trademark section.   It will be a little awkward, as it is not just the addition of a new error condition, but needing to extended/override the current one.

But if this merged as is I wouldn't think any more about it.

**Comment by Jeremy-Kemp** [comment]
Thanks for the feedback @bashbaug, @alycm. I'll swap the change back to the extension spec and update the PR shortly.

**Review by bashbaug — APPROVED** [review]
LGTM - thank you!

**Pull request merged**